[18.0][MIG] queue_job: Migrate to 18.0

QuocDuong1306 opened the pull request:
### Ported PRs
- `#413`
- `#387`
- `#466`
- `#495`
-  `#594`
- `#598`
- `#629`
- `#624`
- `#657`

### Changes

- Cleaned `TODO` tags
- Refactor the `JobSerialized` field on top of `odoo.fields.Json`
   - Result
  
```shell
>>> self.env.cr.execute("SELECT column_name, data_type FROM information_schema.columns WHERE table_name = 'queue_job' AND column_name = 'kwargs';")
>>> self.env.cr.fetchall()
[('kwargs', 'jsonb')]
``` 

  - Example for searching:
  
```shell
>>> self.env.cr.execute("SELECT kwargs->>'a' FROM queue_job WHERE kwargs ? 'a'")
>>> self.env.cr.dictfetchall()
[{'?column?': 'abc'}, {'?column?': 'abc'}]
```

Commit:
Store dependencies

Commit:
Add wait dependencies state

Commit:
Enqueue waiting jobs when parent jobs are done

Commit:
Optimize and make enqueue of waiting jobs more robust

Commit:
Adapt views for state wait_dependencies

Commit:
Add API for Delayables

Commit:
Fix tests failing when test_queue_job is installed

test_queue_job will create channels root.sub and root.sub.sub, so the
tests doing the same thing will fail with a unique constraint error.

Commit:
Add widget to show job dependencies on UI

Commit:
Show the dependency widget in a tab

When displayed in a tab, the widget show the nodes offset, the d3
network is probably confused by the size or visibility of its canvas.
Install a listener on tabs to trigger a fit() on the network.

Commit:
Add documentation on 'base' model public methods

Commit:
Improve loading of dependencies using batch read

Commit:
Use Delayable in DelayableRecordset

OCA-git-bot committed:
queue_job 13.0.1.6.3

OCA-git-bot committed:
queue_job 13.0.1.6.4

eantones committed:
[MIG] queue_job: Migration to 14.0

nilshamerlinck committed:
[ADD] check that queue_job_notify trigger exists

em230418 committed:
[FIX] Flush pending record updates after storing job data

If method in job runs IrModuleModule.button_immediate_install,
environment reset is executed [1]. So env.cr.commit() may silently fail
commiting and stopped job will still be appeared as started,
which is not correct.

This commit fixes described behaviour.

[1] https://github.com/odoo/odoo/blob/aca17a48280e1c1641fde08076197210c926e5f1/odoo/addons/base/models/ir_module.py#L572

nilshamerlinck committed:
[ADD] queue_job: innocuous test job for debugging/monitoring purposes

OCA-git-bot committed:
queue_job 13.0.1.6.5

oca-travis committed:
[UPD] Update queue_job.pot

OCA-git-bot committed:
queue_job 13.0.1.6.6

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-13.0/queue-13.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-13-0/queue-13-0-queue_job/

PCatinean committed:
Added etree._Element to Json encoder/decoder

Commit:
Add method to patch a method to be automatically delayed

This patch method has to be called in ``_register_hook``.

When a method is patched, any call to the method will not directly
execute the method's body, but will instead enqueue a job.

When a ``context_key`` is set when calling ``_patch_job_auto_delay``,
the patched method is automatically delayed only when this key is
``True`` in the caller's context. It is advised to patch the method
with a ``context_key``, because making the automatic delay *in any
case* can produce nasty and unexpected side effects (e.g. another
module calls the method and expects it to be computed before doing
something else, expecting a result, ...).

A typical use case is when a method in a module we don't control is called
synchronously in the middle of another method, and we'd like all the calls
to this method become asynchronous.

It relies on https://github.com/OCA/queue/pull/274 that deprecates the
`@job` decorator.

OCA-git-bot committed:
queue_job 13.0.1.7.0

Commit:
Refactor user_id,model_name,record_ids fields on queue.job

These fields contain the current model and record ids and user id the
job works on.
Storing them as separate fields does not allow to take benefits on the
"JobSerialized" field, which keeps the original user, and later the 'su'
flag. We could store the context there as well (or a cleaned one).

Use computed fields to keep the model_name, record_ids, user_id fields for
backward compatibility. The model_name should stay anyway for the
searches on the UI, and the user_id for the same reason + possibility to
edit the user.

Commit:
Deprecate job decorators

* @job and @related_action are still supported but emit warnings and
  will be removed at some point (code to remove has a TODO with
  :job-no-decorator:)
* the options previously set by the decorators are now stored on
  "queue.job.function" records, which can be explicitly created
  as XML records (and then editable!)
* any method can be used with "with_delay()", if no "queue.job.function"
  exists, default values are used (root channel, default related
  action, ...)

Commit:
Rebind job functions/channels with same name during install

As previously, the job functions and channels were created automatically
by "_register_job" on methods decorated by `@job`, databases have
records without xmlids.

To prevent duplicates (or unique constraint error) in case of channels,
the create method automatically merges the new record with the existing
record. This behavior happens only when installing/upgrading a module,
records created manually behave normally.

If 2 modules adds the same channel or job function, they'll be merged
together with 2 xmlids, on uninstallation of one of the module, the
record will be kept thanks to the second xmlid.

Commit:
Update documentation for deprecation of decorators

Commit:
Add safety to prevent crafting queue.job records

As we can delay a job on any method, and the queue.job model is
accessible from RPC (as any model), prevent to:

* create a queue.job using RPC
* write on protected fields (e.g. method name) using RPC

Admittedly, the risk is low since users need have Queue Job Manager
access to create/write on jobs, but it would allow these users to call
internal methods.

The check is done using a context key that must be equal to a sentinel
object, which is impossible to pass through RPC.

Commit:
Replace job function name by model + method fields

A many2one to select the model + a char for the method name is more
convenient and error-proof than having to type the expected format
correctly.

Commit:
Add missing sudo() when reading queue.job.function

Users shouldn't need access to queue.job.function.
Read config as sudo().

oca-travis committed:
[UPD] Update queue_job.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
queue_job 13.0.3.0.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-13.0/queue-13.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-13-0/queue-13-0-queue_job/

Commit:
Remove warning when @job decorator is used

It makes the runbot builds fail. There is no urgency to remove the
decorator even if we should incentivize the removal because it is
no longer supported, hence the info log. Also, show only the XML snippet
when debug logs are active, because it is too verbose.

OCA-git-bot committed:
queue_job 13.0.3.0.1

OCA-git-bot committed:
queue_job 13.0.3.0.2

oca-travis committed:
[UPD] Update queue_job.pot

OCA-git-bot committed:
queue_job 13.0.3.1.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-13.0/queue-13.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-13-0/queue-13-0-queue_job/

Commit:
Store 'su' flag in records stored in JobSerialized fields

Which means that the record the job is working on, or any record
argument of the job keeps the 'su' flag.

oca-travis committed:
[UPD] Update queue_job.pot

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-13.0/queue-13.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-13-0/queue-13-0-queue_job/

wpichler committed:
Fix missing rollback on retried jobs

When RetryableJobError was raised, any change done by the job was not
rollbacked.

Using `raise` would throw the exception up to the core and rollback, but
we would have a stack trace in the logs for each try. Calling rollback
manually (rollback also clears the env) hide the tracebacks, however,
when the last try fails, the full traceback is still shown in the logs.

Fixes #261

sebastienbeau committed:
[MIG] init V14 repo

- set module uninstallable
- add new .dot file config for v14

sebastienbeau committed:
[IMP] : black, isort, prettier

sebastienbeau committed:
[REF] remove deprecated feature

sebastienbeau committed:
[MIG] finish migration

Commit:
Adapt JobSerialized field to 14.0 changes

The _slots attribute is no longer used. The field properties
are defined as class attributes. See https://github.com/odoo/odoo/pull/48705

Commit:
Remove 13.0 migrations scripts

Commit:
Reorganize models and views in different files

Split one model per file, as advised in the contribution guide.

Commit:
Fix error with undefined related action kwargs

Commit:
Add import shortcut for identity_exact

Now that the `@job` decorater is deleted, `identity_exact` is the only
thing we may need to import from "job.py".
Being able to import it from "odoo.addons.queue_job" is nicer than
from "odoo.addons.queue_job.job".

Commit:
Improve error handling on invalid job function dicts

oca-travis committed:
[UPD] Update queue_job.pot

OCA-git-bot committed:
[UPD] README.rst

Commit:
Fix required sudo in Job

Following changes of https://github.com/OCA/queue/pull/281
The initial sudo() is lost when we call "with_env()" with a False su
flag. Ensure the read job.record keeps a su flag.

OCA-git-bot committed:
queue_job 14.0.1.0.1

nilshamerlinck committed:
[FIX] run doctests as part of standard tests

Commit:
Remove initial create notification and follower

Everytime a job is created, a mail.message "Queue Job created" is
created. This is useless, as we already have the creation date and user,
and nobody will ever want to receive a notification for a created job
anyway.

Removing the on creation auto-subscription of the user that created the
job makes sense as well since we automatically subscribe the queue job
managers for failures. Add the owner of the jobs to the followers on
failures only as well.

It allows to remove a lot of insertions of records (and of deletions
when autovacuuming jobs).

Commit:
Change technical fields to read-only

These fields should not be changed by users.

Commit:
Optimize queue.job creation

Several fields on queue.job are initialized using computed fields,
then never changed again. On creation of a queue.job record, we'll have
an initial INSERT + at least one following UPDATE for the computed
fields.

Replace all the stored computed fields by a raw initialization of the
values in `Job.store()` when the job is created, so we have only a
single INSERT.

Highlights:

* as channel is no longer a compute/inverse field, override_channel is
  useless, I dropped it (actually the same value was stored in both
  channel and override_channel as the channel field was stored)
* one functional diff is that now, when changing a channel on a
  job.function, the channel is no longer synchronized on existing jobs,
  it will be applied only on new jobs: actually this is an improvement,
  because it was impossible to change the channel of a job function
  in a large queue_job table as it meant writing on all the done/started
  jobs
* searching the queue.job.function is now cached, as each job using the
  same will run a query on queue_job_function

simahawk committed:
queue_job: add exec time to view some stats

Commit:
Add model in search view / group by

Commit:
Fix date_done set when state changes back to pending

When Job.date_done has been set, for instance because the job has been
performed, if the job is set back to pending (e.g. a RetryableJobError
is raised), the date_done is kept.

florian-dacosta committed:
Vacuum done jobs in batch
Huge amount of jobs to delete may take a lot of time and the cron may crash, because of timeout for instance
This issue would make the cron permanently run and permanently failing... Deleting batch allow to avoid this issue
if the cron fails, it still will have delete some of the job history

OCA-git-bot committed:
queue_job 14.0.1.0.2

codingforfun committed:
[FIX] Respect `cron.user_id` for manually triggered queue jobs

PieterPaulussen committed:
[IMP] Store job execution information + wizard to terminate a job + test

simahawk committed:
queue_job: close buffer when done

simahawk committed:
queue_job: store exception name and message

simahawk committed:
queue_job: improve filtering and grouping

simahawk committed:
queue_job: add hook to customize stored values

simahawk committed:
queue_job: migration step to store exception data

Commit:
[ADD] Apply suggestions from code review

Co-authored-by: Simone Orsi <simahawk@users.noreply.github.com>

PieterPaulussen committed:
[FIX] add test tags to DocTestCase since failures are not registered correctly with OdooDocTestCase

StefanRijnhart committed:
[FIX] Patch Python 3.8 class teardown API to doctest

PieterPaulussen committed:
[FIX] Ouput a clear log statement when starting the doctest

PieterPaulussen committed:
[FIX] Restore use of relative import

davejames committed:
[FIX] replace deprecated use of logging.warn

OCA-git-bot committed:
queue_job 14.0.1.0.3

oca-travis committed:
[UPD] Update queue_job.pot

OCA-git-bot committed:
queue_job 14.0.1.1.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-14.0/queue-14.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-14-0/queue-14-0-queue_job/

guewen committed:
Revert "[14.0] [IMP] Store job execution information + wizard to terminate a job + test"

oca-travis committed:
[UPD] Update queue_job.pot

OCA-git-bot committed:
queue_job 14.0.1.2.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-14.0/queue-14.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-14-0/queue-14-0-queue_job/

hbrunn committed:
[IMP] queue_job: Add cancelled state to queue.job

PieterPaulussen committed:
[ADD] [14.0] randomize retry interval

hbrunn committed:
[IMP] queue_job: tests for wizards

guewen committed:
Add documentation

guewen committed:
Add a graph UUID

A graph of jobs always share the same graph_uuid, which can be
used to group jobs, but is also a faster way to find all the jobs
of the graph. Then we can use the dependencies field to build the
whole graph from the pre-selection of jobs.

guewen committed:
Hide some technical fields

guewen committed:
Ignore requeues on dependency jobs waiting on parent jobs

Jobs waiting that their dependencies are executed cannot be requeued.
They have to keep waiting their turn.

guewen committed:
Fix lint

OCA-git-bot committed:
queue_job 14.0.1.3.0

guewen committed:
Update vis-network js

guewen committed:
Improve display of jobs graph widget

guewen committed:
Add powerful context manager for running tests on jobs

* mock_jobs context manager with assert methods to check enqueued jobs
* jobs are not stored in database, yet they can be inspected and performed
  during the test
* the context manager works with graphs enqueued by delayable and with with_delay()
* direct execution of jobs works with graphs by executing them synchronously
  following their topological sort

guewen committed:
Set graph_uuid only once in DelayableGraph

The previous implementation was incomplete, as it was setting
the graph_uuid when `Job.add_depends()` was called. It had the
advantage of being correct when (for e.g. a test) manually created
a Job and called `add_depends()`, but it could not work when in
some situations involving a group where one of the job has no job
depending on it.

Also, remove ``Job.add_reverse_depends()`` which is not used.

guewen committed:
Add docstrings on the new delayable classes

guewen committed:
Fix display on exec_time on tree view as seconds

The float_time widget shows hours seconds, we only want seconds.
The widget had been removed on the form view, but not on the tree view.

etobella committed:
[FIX] queue_job: Migrations raising errors with OpenUpgrade

guewen committed:
Add option to generate a graph in create_test_job controller

guewen committed:
Improve graph widget

* Fix a js bug when there is no graph.
* Use a single read group for multiple jobs

guewen committed:
Add a smart button to open all the jobs of a graph

guewen committed:
ix duplicate label

guewen committed:
Fix graph widget now showing title as HTML

From https://visjs.github.io/vis-network/examples/network/other/html-in-titles.html

guewen committed:
Improve graph widget performance

When showing large graphs, like 300 jobs (not really a real life case maybe),
the graph takes a dozen of seconds to render. Disabling the stabilization
helps a lot.

legalsylvain committed:
[DOC] describe how to write queue.job.function in case of function defined in abstract model

guewen committed:
Escape strings passed to the graph js widget

OCA-git-bot committed:
queue_job 14.0.1.3.1

acsonefho committed:
[14.0][IMP] queue_job current company
Use the current company to trigger the job (+ add related tests)

[14.0][FIX] queu_job: allowed_company_ids => use with_company(...)
Fill allowed_company_ids from context with the job's company instead of every allowed companies of the user.
Because most of the time, a job is related to only one company. And adding every allowed companies of the user into the context may load some unexpected records (during search for example). Because standards ir.rule use ['|',('company_id','=',False),('company_id', 'in', company_ids)] and this 'company_ids' is filled with every allowed companies from the context.

oca-travis committed:
[UPD] Update queue_job.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-14.0/queue-14.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-14-0/queue-14-0-queue_job/

romi477 committed:
[IMP] queue_job: black, isort, prettier

romi477 committed:
[MIG] queue_job: Migration to 15.0

eLBati committed:
FIX queue_job avoiding AccessDenied when loading the module in an instance with list_db = False
The scenario is a multi database environment where you can't set db_name as the databases change frequently and you use dbfilter to allow users to access their DB according to domain

guewen committed:
Set modules as uninstallable

guewen committed:
Set modules queue_job and test_queue_job as installable

guewen committed:
queue_job: use sudo() instead of with_user(SUPERUSER_ID)

guewen committed:
Change pylint disable instruction with human-readable code

Commit:
[UPD] Update queue_job.pot

OCA-git-bot committed:
[UPD] README.rst

fernandahf committed:
[IMP] queue_job: use a widget in eta field in queue job tree view

em230418 committed:
queue_job: fix typo in USAGE.rst

dzungtran89 committed:
[IMP] queue_job: black, isort, prettier

dzungtran89 committed:
Forward migration scripts from #309 #328

guewen committed:
Add documentation about handling of failures in a graph

guewen committed:
Fix equality of enqueued jobs

The assert fails to show the method's details if we store the __func__ in JobCall.
Anyway, it was hackish to change the content of "method".

guewen committed:
Use a python3.6 compatible data class

guewen committed:
Rename mock_jobs() to trap_jobs()

The fact that mocks are used is an implementation detail, and is actually
hidden to the end user of the test API.

guewen committed:
Rename dependency method done() to on_done()

on_done() is maybe a better choice, it is less ambiguous than then() (clearer
on the fact it happens on done()) and done() (less confusion with set_done()).
I'll do the renaming soon, unless you have other suggestions.

Ref: https://github.com/OCA/queue/pull/154#issuecomment-1025127060

moylop260 committed:
[FIX] queue_job: runner - filedescriptor out of range in select

Use the most efficient Selector implementation available on the current platform

Odoo supports only SelectSelector but it is a little obsolete

python >= 3.4 supports a new high-level library Selectors:
 - https://docs.python.org/es/3/library/selectors.html

It could to auto-choose the following ones:
 - SelectSelector
 - PollSelector
 - EpollSelector
 - DevpollSelector
 - KqueueSelector

Using the DefaultSelector class the most efficient implementation available on the current platform will be use:
 - https://docs.python.org/3/library/selectors.html#selectors.DefaultSelector

It helps to support better the resources of the system

Using SelectSelector you are not able to run workers >=255

If you set `ulimit -n 10240` and run `odoo-bin --workers=255`
the following error is raised:

    Traceback (most recent call last):
    File "odoo/service/server.py", line 926, in run
        self.sleep()
    File "odoo/service/server.py", line 852, in sleep
        sel.select(self.beat)
    File "python3.8/lib/python3.8/selectors.py", line 323, in select
        r, w, _ = self._select(self._readers, self._writers, [], timeout)
    ValueError: filedescriptor out of range in select()

But using PollSelector it is not reproduced even using more workers

Most of platform supports PollSelector but using DefaultSelector we can be sure
that even too old system are supported too

And using this High-level library will allow to use the future new improvements

e.g. Epoll has better performance improvements

More info about:
 - https://devarea.com/linux-io-multiplexing-select-vs-poll-vs-epoll
 - https://github.com/redis/redis-py/issues/486
 - https://github.com/odoo/odoo/pull/84684

sbidoul committed:
add jobrunner cli

simahawk committed:
queue_job: use parent channel if configured

Use case:

* you have a root channel per scope/app (eg: root.edi)
* you have several sub channels (eg: root.edi.ubl.sales,
root.edi.gs1.delivery)
* you want to configure capacity only for the main channel "root.edi"

Before this change, the channel manager falls back on root channel.
However, if you have a specific parent channel configured
it sounds a good idea to use it.

simahawk committed:
queue_job: update contributors

Contributing since a while... :)

simahawk committed:
queue_job: fix pylint duplicate-xml-fields

etobella committed:
[IMP] queue_job: Recover gracefully when the CPU limit is reached

OCA-git-bot committed:
queue_job 15.0.1.0.1

guewen committed:
Add storage of context in jobs recordsets

guewen committed:
Disable storing of context in jobs recordsets by default

It can still be activated by overriding a method.

guewen committed:
Fix cursor already closed error on retryable errors

When a job is postponed because of a retryable error, as the error
is not re-raised, we reach the new method

    self._enqueue_dependent_jobs(env, job)

With a closed cursor. Return early in case of postponing, as there is
no chance dependent jobs can become pending anyway.

victoralmau committed:
[FIX] queue_job: Add base_sparse_field dependency to prevent errors in test if addon not loaded.
TT35938

yajo committed:
Added translation using Weblate (Spanish)

yajo committed:
Translated using Weblate (Spanish)

Currently translated at 3.7% (5 of 135 strings)

Translation: queue-15.0/queue-15.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-15-0/queue-15-0-queue_job/es/

yajo committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (135 of 135 strings)

Translation: queue-15.0/queue-15.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-15-0/queue-15-0-queue_job/es/

Commit:
[IMP] queue_job: Create job with saved context

dong-z committed:
Translated using Weblate (Chinese (Simplified))

Currently translated at 83.2% (114 of 137 strings)

Translation: queue-15.0/queue-15.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-15-0/queue-15-0-queue_job/zh_CN/

victoralmau committed:
[FIX] queue_job: Update documentation to remove unnecessary addon (base_sparse_field) from server_wide_modules.
https://github.com/OCA/queue/pull/443#discussion_r946913070

OCA-git-bot committed:
queue_job 15.0.1.0.2

ivantodorovich committed:
[IMP] `queue_job`: bump app icon style

baimont committed:
[MIG] queue_job: Migration to 16.0

sbidoul committed:
Initialize 16.0

baimont committed:
[FIX] import mock from unittest

Commit:
[UPD] Update queue_job.pot

OCA-git-bot committed:
[UPD] README.rst

Commit:
[UPD] Update queue_job.pot

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-16.0/queue-16.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-16-0/queue-16-0-queue_job/

sbidoul committed:
Revert "[14.0][IMP] queue_job current company"

This reverts commit 309eb5fe55cb5c32ab4f075e619ccf81d46c74f1.

baimont committed:
queue_job - activate the base context keys

Co-authored-by: Stéphane Bidoul <stephane.bidoul@acsone.eu>

guewen committed:
Migrate queue job graph/dependencies to 15.0

guewen committed:
Apply pre-commit on migration of jobs graph

guewen committed:
Fix things required by odoo 14.0 or python 3.9

* Renaming the TestXxx models in test_queue_job is because of pytest that
  complains, confusing them for test classes
* logger.warn is deprecated in favor of logger.warning
* Add invalidation of cache in SQL update that enqueue dependent jobs
* Flush in a test, the same is done currently in the controller
* Weird one: we can no longer compare the bound methods of the models,
  compare their __func__ which is equal.

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
queue_job 16.0.1.0.1

Commit:
[UPD] Update queue_job.pot

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-15.0/queue-15.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-15-0/queue-15-0-queue_job/

OCA-git-bot committed:
queue_job 16.0.1.0.2

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
queue_job 16.0.1.1.0

repodevs committed:
[IMP] queue_job: add configuration options for
jobrunner_db_user, jobrunner_db_password

marylla committed:
Translated using Weblate (German)

Currently translated at 86.3% (114 of 132 strings)

Translation: queue-16.0/queue-16.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-16-0/queue-16-0-queue_job/de/

Commit:
[UPD] Update queue_job.pot

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-15.0/queue-15.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-15-0/queue-15-0-queue_job/

lmignon committed:
[MIG] queue_job, test_queue_job: Forward port queue from 15.0

This forward port takes all the changes from 15.0 not yet into 16.0
up to https://github.com/OCA/queue/commit/173b6ca7026dc8042cc8e750767c3267c67f4f43

This FP contains the new queue dependencies functionality.
The dedicated widget used to display the dependecies graph on the job
form is also migrated to owl

lmignon committed:
[FIX] queue_job: fix invalid po files

Commit:
[UPD] Update queue_job.pot

OCA-git-bot committed:
queue_job 16.0.2.0.0

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-16.0/queue-16.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-16-0/queue-16-0-queue_job/

etobella committed:
[FIX] queue_job: delayable representation failed

StefanRijnhart committed:
[FIX] queue_job: UI glitches in 16.0 migration

Commit:
[UPD] Update queue_job.pot

OCA-git-bot committed:
queue_job 16.0.2.1.0

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-16.0/queue-16.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-16-0/queue-16-0-queue_job/

hparfr committed:
IMP queue_job: add cancelled filter

hugosantosred committed:
[FIX] Patch Request._get_session_and_dbname to get db from URL

In servers with multiple databases db is not captured from URL when
processing jobs in /queue_job/runjob

With this patch db is set from the db param in URL when calling
the runjob endpoint

Fixes #503

sanchonuria committed:
[14.0][FIX] queue_job: fix read

etobella committed:
[FIX] queue_job: Commit was removed, but that could be a problem if you generated too much records

FranciscoFactorLibre committed:
Translated using Weblate (Spanish)

Currently translated at 82.0% (128 of 156 strings)

Translation: queue-16.0/queue-16.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-16-0/queue-16-0-queue_job/es/

OCA-git-bot committed:
queue_job 16.0.2.2.0

simahawk committed:
queue_job: unify no delay option

Current situation:

* multiple keys for no good reason
* half baked: not all of them used everywhere
* no centralization
* poor naming

With this change we'll have:

* 1 and only one key to disable via ctx: ``queue_job__no_delay``
* 1 and only one key to disable via os env: ``QUEUE_JOB__NO_DELAY``
* backward compatibility with deprecation for old keys

OCA-git-bot committed:
queue_job 16.0.2.2.1

OCA-git-bot committed:
queue_job 16.0.2.3.0

OCA-git-bot committed:
[UPD] README.rst

sbidoul committed:
Handle cancelled state in job runner

OCA-git-bot committed:
queue_job 16.0.2.3.1

guewen committed:
Fix trap_jobs() recordset comparison

The trap_jobs() test helper accepts any model/recordset as long
as the job function is the same, which is not enough to verify that
the job has been delayed on the expected recordset or model.

Tested job functions are "bound methods" so we can compare their `__self__`.

Example of assertion that would be valid:

  partner = self.env["res.partner"].create({"name": "foo"})
  with trap_jobs() as trap:
    partner.with_delay().do_something(42)
    trap.assert_enqueued_job(
        self.env["res.partner"].do_something,
        args=(42,)
    )
    trap.assert_jobs_count(1, only=self.env["res.partner"].do_something)

Now, it would fail with an error message like:

E           AssertionError: Job <res.partner()>.do_something(42) with properties (channel=root.test, description=Do something, eta=15, identity_key=<function identity_exact at 0x7f01bfb54cb0>, max_retries=1, priority=15) was not enqueued.
E           Actual enqueued jobs:
E            * <res.partner(64,)>.do_something(42) with properties (priority=15, max_retries=1, eta=15, description=Do something, channel=root.test, identity_key=<function identity_exact at 0x7f01bfb54cb0>)

ramiadavid committed:
Translated using Weblate (Spanish)

Currently translated at 98.0% (153 of 156 strings)

Translation: queue-16.0/queue-16.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-16-0/queue-16-0-queue_job/es/

ramiadavid committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (156 of 156 strings)

Translation: queue-16.0/queue-16.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-16-0/queue-16-0-queue_job/es/

OCA-git-bot committed:
queue_job 16.0.2.3.2

OCA-git-bot committed:
queue_job 16.0.2.4.0

Commit:
[IMP] queue_job: identity_key enhancements

1. In production, a job which is waiting dependencies or which
has started, but not completed, should not be repeated if
the identity_key matches.

2. In tests, the mock queue handler is now enhanced to allow
better mimicking of the identity_key blocks from production.

3. In tests, the mock queue handler now clears the enqueued jobs
after performing them, to better reproduce what a production
environment would do.

OCA-git-bot committed:
queue_job 16.0.2.5.0

Commit:
[UPD] Update queue_job.pot

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-16.0/queue-16.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-16-0/queue-16-0-queue_job/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (156 of 156 strings)

Translation: queue-16.0/queue-16.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-16-0/queue-16-0-queue_job/es/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (156 of 156 strings)

Translation: queue-16.0/queue-16.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-16-0/queue-16-0-queue_job/es/

Commit:
[UPD] Update queue_job.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-16.0/queue-16.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-16-0/queue-16-0-queue_job/

MiquelRForgeFlow committed:
[IMP] queue_job: track error in chatter

simahawk committed:
queue_job: job_record_with_same_identity_key ignore STARTED

Good explanation from Guewen:

Depending on the use case, we should or should not include.
As part of my current work, I use Sidekiq daily.
They have a similar feature, but you can set a per-job parameter unique_until with options:
start (that would mean up to ENQUEUED here) or success (that would be up to STARTED here).

Think about this use case: a job refreshes a cache. Data have changed, we create a pending job. Data change again, no new job because the job is still pending.
Job starts. Data change while the job is running. In this very case, we'd like to enqueue a new job otherwise the cache will be outdated.

I reckon that both cases are valid, but I fear adding this state in the domain may, silently and in subtle ways, existing behaviors.

pcastelovigo committed:
queue job avoiding AccessDenied when loading the module in an instance with list_db=false

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
[BOT] post-merge updates

simahawk committed:
queue_job: split identity_key hasher to ease reuse

You can now create your own identity exact matcher using the default hasher as a base.

nguyenminhchien committed:
[IMP] queue_job: pre-commit auto fixes

nguyenminhchien committed:
[IMP] queue_job: pre-commit auto fixes

nguyenminhchien committed:
[MIG] queue_job: Migration to 17.0

Commit:
[UPD] Update queue_job.pot

OCA-git-bot committed:
[BOT] post-merge updates

InigoGR committed:
queue_job: fix retry format with tuple values

#348 added support for randomized retry intervals.
Configuration of randomized retry intervals is not possible due to the formatting checks not being updated.
This should fix it.

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: queue-17.0/queue-17.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-17-0/queue-17-0-queue_job/

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 68.1% (105 of 154 strings)

Translation: queue-17.0/queue-17.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-17-0/queue-17-0-queue_job/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (154 of 154 strings)

Translation: queue-17.0/queue-17.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-17-0/queue-17-0-queue_job/it/

sebalix committed:
queue_job: fix partial index to add 'wait_dependencies' state

AlexanderPashuk committed:
Remove doctest from get_db_names method to avoid issues with Odoo.sh

sebalix committed:
queue_job: triggers stored computed fields before calling 'set_done()'

So the time required to compute such fields by the ORM is taken into
account when the 'date_done' and 'exec_time' values are set on the job.

OCA-git-bot committed:
[BOT] post-merge updates

pedrobaeza committed:
[FIX] *queue_job: Pre-commit fixes

OCA-git-bot committed:
[BOT] post-merge updates

fd-oncodna committed:
[fix] queue_job: JobsTrap.perform_enqueued_jobs
Closes #651

OCA-git-bot committed:
[BOT] post-merge updates

paradoxxxzero committed:
[IMP] queue_job: Display warning before displaying big dependency graphs

xtanuiha committed:
Translated using Weblate (Chinese (Simplified) (zh_CN))

Currently translated at 100.0% (154 of 154 strings)

Translation: queue-17.0/queue-17.0-queue_job
Translate-URL: https://translation.odoo-community.org/projects/queue-17-0/queue-17-0-queue_job/zh_CN/

QuocDuong1306 committed:
[IMP] queue_job: Cancel child jobs when the parent is cancelled

florentx committed:
[FIX] queue_job: typo

florentx committed:
[IMP] queue_job: add filter on Date Created

QuocDuong1306 committed:
[MIG] queue_job: Migration to 18.0

nilshamerlinck commented on queue_job/controllers/main.py:
```diff
@@ -0,0 +1,297 @@
+
+                _logger.debug("%s OperationalError, postponed", job)
+                raise RetryableJobError(
+                    tools.ustr(err.pgerror, errors="replace"), seconds=PG_RETRY
```
`tools.ustr` is deprecated, see `https://github.com/odoo/odoo/pull/173811`

thienvh332 commented on queue_job/migrations/14.0.3.5.2/pre-migration.py:
```diff
@@ -0,0 +1,10 @@
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html)
+
+from odoo.tools.sql import table_exists
+
+
+def migrate(cr, version):
+    if table_exists(cr, "queue_job"):
+        # Drop index 'queue_job_identity_key_state_partial_index',
+        # it will be recreated during the update
+        cr.execute("DROP INDEX IF EXISTS queue_job_identity_key_state_partial_index;")
```
Remove it

thienvh332 commented on queue_job/utils.py:
```diff
@@ -0,0 +1,40 @@
+
+    :param env: `odoo.api.Environment` instance
+    """
+    # TODO: drop in v17
+    if os.getenv("TEST_QUEUE_JOB_NO_DELAY"):
+        _logger.warning(
+            "`TEST_QUEUE_JOB_NO_DELAY`  env var found. NO JOB scheduled. "
+            "Note that this key is deprecated: please use `QUEUE_JOB__NO_DELAY`"
+        )
+        return True
```
We should delete it. It's out of date.

thienvh332 commented on queue_job/utils.py:
```diff
@@ -0,0 +1,40 @@
+        _logger.warning("`QUEUE_JOB__NO_DELAY` env var found. NO JOB scheduled.")
+        return True
+
+    # TODO: drop in v17
+    deprecated_keys = ("_job_force_sync", "test_queue_job_no_delay")
+    for key in deprecated_keys:
+        if env.context.get(key):
+            _logger.warning(
+                "`%s` ctx key found. NO JOB scheduled. "
+                "Note that this key is deprecated: please use `queue_job__no_delay`",
+                key,
+            )
+            return True
```
Same here

thienvh332 commented on queue_job/job.py:
```diff
@@ -0,0 +1,906 @@
+        )
+        return existing
+
+    # TODO to deprecate (not called anymore)
+    @classmethod
+    def enqueue(
```
Check and delete it if no longer used.

thienvh332 commented on queue_job/job.py:
```diff
@@ -0,0 +1,906 @@
+        )
+        return new_job._enqueue_job()
+
+    # TODO to deprecate (not called anymore)
+    def _enqueue_job(self):
```
Check and delete it if no longer used.

thienvh332 commented on queue_job/job.py:
```diff
@@ -0,0 +1,906 @@
+        )
+        return self
+
+    @staticmethod
+    def db_record_from_uuid(env, job_uuid):
+        # TODO remove in 15.0 or 16.0
+        _logger.debug("deprecated, use 'db_records_from_uuids")
+        return Job.db_records_from_uuids(env, [job_uuid])
```
Check and delete it if no longer used.

thienvh332 commented:
Have you added the content of [this commit](https://github.com/OCA/queue/pull/657/commits/f0511aa90ed652e6d7ae76ee40cd5f098047ab60)?

QuocDuong1306 committed:
[IMP] queue_job: Cleaned TODO tags

QuocDuong1306 commented:
Hi @thienvh332 , actually, that commit was used for the old version. Btw I'll take a chance to pick and migrate it. While waiting, could you have a look at the updated code?

QuocDuong1306 commented:
Hi a @thienvh332 , updated the code

thienvh332 commented on queue_job/static/src/views/fields/job_direct_graph/job_direct_graph.scss:
```diff
@@ -7,12 +7,4 @@
-        align-items: center;
-        justify-content: center;
-    }
-}
```
I see you are still using it. Are you sure deleting this won't affect anything?

QuocDuong1306 replied:
Hi a, I replaced those CSS with explicit classes.

![image](https://github.com/user-attachments/assets/a725dad3-ec6b-440a-b314-4d90e74829ad)

thienvh332 reviewed (approved):
LGTM

QuocDuong1306 committed:
[IMP] queue_job: use jsonb column for JobSerialized

- Refactor the JobSerialized field on top of `odoo.fields.Json` to use jsonb column

QuocDuong1306 commented:
Hello guys, new commit was added for this one
> It would be nice to refactor the JobSerialized field on top of `odoo.fields.Json` to use jsonb column and be able to search jobs using the more advanced JSONB features.

nilshamerlinck commented on queue_job/models/queue_job.py:
```diff
@@ -0,0 +1,504 @@
+
+    model_name = fields.Char(string="Model", readonly=True)
+    method_name = fields.Char(readonly=True)
+    # record_ids field is only for backward compatibility (e.g. used in related
+    # actions), can be removed (replaced by "records") in 14.0
+    record_ids = JobSerialized(compute="_compute_record_ids", base_type=list)
```
check comment

QuocDuong1306 replied:
Thank Nils, I updated the code

nilshamerlinck commented on queue_job/fields.py:
```diff
@@ -0,0 +1,126 @@
+        return default_json
+
+    def convert_to_column(self, value, record, values=None, validate=True):
+        return self.convert_to_cache(value, record, validate=validate)
```
shouldn't we return a PsycopgJson?

QuocDuong1306 replied:
Hi nils, updated code, also took a chance to improve `convert_to_export` (change the encoder)

nilshamerlinck commented:
thanks

can you just remove this part from commit: `- And be able to search jobs using the more advanced JSONB features`

because we don't do anything for that for now; fyi I asked for Simone's opinion about it in card:

![image](https://github.com/user-attachments/assets/e604572f-edb1-4424-87f1-a6803b2ab656)
